python/hail/expr/expression.py

```diff
@@ -3546,77 +3577,82 @@ def analyze(expr, expected_indices, aggregation_axes=set()):
 
     if source is not None and source is not expected_source:
         errors.append(
-            ExpressionException('Expected an expression from source {}, found expression derived from {}'.format(
-                expected_source, source
+            ExpressionException('{} expects an expression from source {}, found expression derived from {}'.format(
+                caller, expected_source, source
             )))
 
-    # TODO: use ast.search to find the references to bad-indexed exprs
-
-    # check for stray indices
+    # check for stray indices by subtracting expected axes from observed
     unexpected_axes = axes - expected_axes
+
     if unexpected_axes:
+        # one or more out-of-scope fields
+        assert len(refs) > 0
+        bad_refs = []
+        for name, inds in refs:
+            bad_axes = inds.axes.intersection(unexpected_axes)
+            if bad_axes:
+                bad_refs.append((name, inds))
+
+        assert len(bad_refs) > 0
         errors.append(ExpressionException(
-            'out-of-scope expression: expected an expression indexed by {}, found indices {}'.format(
-                list(expected_axes),
-                list(indices.axes)
+            "scope violation: '{caller}' expects an expression indexed by {expected}"
+            "\n    Found indices {axes}, with unexpected indices {stray}. Invalid fields:{fields}{agg}".format(
+                caller=caller,
+                expected=list(expected_axes),
+                axes=list(indices.axes),
+                stray=list(unexpected_axes),
+                fields=''.join("\n        '{}' (indices {})".format(name, list(inds.axes)) for name, inds in bad_refs),
+                agg='' if (unexpected_axes - aggregation_axes) else
```

Comment:
why use this condition for determining whether to print the "supports aggregation over" message?

Reply:
This condition means that all stray axes are available for aggregation. It's possible you use GT in a `select_global` or something that doesn't support aggregation. In that case we don't want the message.